### ISSUE TITLE
What makes a good R package?

### ISSUE THREAD

**Issue by karthik** [ropensci/unconf#33]
At rOpenSci we are in the midst of formalizing our standards and guidelines for what we would consider 'acceptable' package for our suite (beyond the actual fit for our scope). We are particularly keen on hearing from the community about how to improve on our checklist and other things we can check for to lift the overall quality of scientific R packages that eventually make it on CRAN. 

A common complaint about R and CRAN packages is that there are several crappy packages, many without any tests. Given that we are playing an increasingly larger role in this space, how can we help effect some change?

We began a first draft here: https://github.com/ropensci/onboarding#-ropensci-onboarding-

Feedback welcome on that repo, but also a small discussion in person would be great.

**Comment by jennybc** [comment]
Something R gets beat up about constantly is inconsistency around word delimiters. In function names, function arguments, etc. A small step in the right direction would be to articulate one or more standards and build automated checks for them, a la `R CMD check`. Less easy to automate and more in the spirit of _some_ of what CRAN gatekeepers do re: DESCRIPTION files is to make sure that … function names are verbs, the most specific/evocative word comes first, etc etc. [This kind of stuff](http://blog.stephenwolfram.com/2010/10/the-poetry-of-function-naming/). If we ever developed a framework for package peer review, perhaps coupled to a repository or badging system, this is the sort of thing reviewers could help authors spot and improve.

**Comment by gavinsimpson** [comment]
Naming conventions aside (do you _really_ want to open that particular can of worms?),  good packages are ones that have consistent interfaces _internally_; so the functions work similarly and interoperably with other functions in the package. Good packages are _well_ documented with useful examples that illustrate the salient usage examples; that clearly explain what the arguments mean (not what the author thinks everyone knows they mean), that contain links to literature etc. Good packages have a vignette and unit tests. Good packages are ones that have been shown love and attention by their authors'.

I'm not sure how much of this you can formalise; you can expect a vignette or tests and authors may supply them but if they aren't very good then they don't help. Yet how to do formalise that in guidelines without a proper review process like we have for papers.

Don't forget, CRAN has a pretty lengthy (and infamous) checklist of things developers need to do (hoops to jump through) to get packages on CRAN and there is plenty of crap(pier) packages there. What they can't check without formal review are exactly those things that make good packages stand out from the mediocre.

**Comment by Ironholds** [comment]
Yeah; I wish there was a way of participating in CRAN as a reviewer to make this more workable :/. But I think if the threshold is "here is what we expect rOpenSci packages to have", reviewing directly is more possible.

**Comment by gavinsimpson** [comment]
@Ironholds In a round-about way that was my point. Guidelines won't work or achieve the desired aim, but a formal, open, review process _could_.

There have been attempts (e.g. Crantastic) to have some form of package review but they haven't been successful because, well, people. I think for rOpenSci however there is sufficient a community that could make this work.

**Comment by Ironholds** [comment]
Gotcha. Then a big +1 to that! I'd love a guidelines + reviewers combo, and consider me signed up as the first reviewer.

**Comment by drisso** [comment]
I think a good place to look at is Bioconductor. They have a set of requirements (they call them guidelines but many sections start with "a package must ..."): http://www.bioconductor.org/developers/package-guidelines/

Extrapolating from their list, I think a minimum set of good features is:
- a package vignette + documentation for all functions
- unit tests
- package maintainer responsibilities (many CRAN packages are "abandoned")
- version numbering conventions

It's also worth checking their BiocCheck which complements the usual R CMD check.

PS. They do have a formal review process for new package submissions.

**Comment by kplaney** [comment]
Yes ditto on the Bioconductor package submission process. It's still not perfect, but you have several back-and-forth iterations with a reviewer before you can publish your package. The reviewers expect a variety of examples for all your functions and full-fledged vignette, and baseline unit tests or if statements.  Right now they don't strictly require using R's Runit, but they do recommend it when you check your package (you do R CMD Check, and then also run Bioconductor's own "check".)  

From personal experience, I still don't think Bioconductor requires quite enough unit tests, especially ones tailored with error messages that are helpful to the user. While not sophisticated, I try and put in lots of if statements that output warnings or errors to users with specific hypotheses about what went wrong (i.e. what was probably wrong with their data format, or if an unexpected error/bug occurred that I know shouldn't happen - if my code was correct...) It's also helpful for me if it's a rather lengthy statistical algorithm! :)

**Comment by HenrikBengtsson** [comment]
Regarding @jennybc's note on naming and coding conventions: It's obvious that there are many flavors out there, and then there's the random ones.  One could imagine trying to formalize it by identify a few of the common styles and provide an option (optional) to specify which one the package follows (e.g. DESCRIPTION field 'CodingStyle: BlaBlav2') and leverage `R CMD check` to test/give feedback relative to the target style.
